core/corehttp/gateway_handler.go

```diff
@@ -92,16 +92,24 @@ func (i *gatewayHandler) getOrHeadHandler(w http.ResponseWriter, r *http.Request
 
 	urlPath := r.URL.Path
 
+	// If the gateway is behind a reverse proxy and mounted at a sub-path,
+	// the prefix header can be set to signal this sub-path.
+	// It will be prepended to links in directory listings and the index.html redirect.
+	prefix := ""
+	if prefixHdr := r.Header["X-Ipfs-Gateway-Prefix"]; len(prefixHdr) > 0 {
+		log.Debugf("X-Ipfs-Gateway-Prefix: %s", prefixHdr[0])
+		prefix = prefixHdr[0]
+	}
+
 	// IPNSHostnameOption might have constructed an IPNS path using the Host header.
 	// In this case, we need the original path for constructing redirects
 	// and links that match the requested URL.
 	// For example, http://example.net would become /ipns/example.net, and
 	// the redirects and links would end up as http://example.net/ipns/example.net
-	originalUrlPath := urlPath
+	originalUrlPath := prefix + urlPath
 	ipnsHostname := false
-	hdr := r.Header["X-IPNS-Original-Path"]
-	if len(hdr) > 0 {
-		originalUrlPath = hdr[0]
+	if hdr := r.Header["X-Ipns-Original-Path"]; len(hdr) > 0 {
```

Comment:
net/http seems to convert response header keys to camel-case. I passed `X-IPFS-Gateway-Prefix` from nginx, and it arrived in the gateway handler as `X-Ipfs-Gateway-Prefix`. I only changed `X-IPNS-Original-Path` too for the sake of consistency and avoiding happy accidents.